File: dsa-problems/Bitwise/solutions/python/BIT-002-two-unique-with-triples-mask.py
import sys

def two_unique_with_triples_mask(a: list[int], M: int) -> list[int]:
    # 1. Find splitting bit index (must be in M and distinguish the two uniques)
    split_bit = -1
    for i in range(31):
        if not ((M >> i) & 1):
            continue

        count = 0
        for x in a:
            if (x >> i) & 1:
                count += 1

        if count % 3 == 1:
            split_bit = i
            break

    # If no split bit found in M, find any distinguishing bit
    if split_bit == -1:
        for i in range(31):
            count = 0
            for x in a:
                if (x >> i) & 1:
                    count += 1
            if count % 3 == 1:
                split_bit = i
                break

    # 2. Reconstruct (now split_bit is guaranteed to be valid)
    num1, num2 = 0, 0
    for i in range(31):
        c1, c2 = 0, 0
        for x in a:
            bit_val = (x >> i) & 1
            if (x >> split_bit) & 1:
                c2 += bit_val
            else:
                c1 += bit_val

        if c1 % 3 == 1: num1 |= (1 << i)
        if c2 % 3 == 1: num2 |= (1 << i)

    return sorted([num1, num2])

def main():
    input = sys.stdin.read
    data = input().split()
    if not data: return

    ptr = 0
    n = int(data[ptr]); ptr += 1
    a = []
    for _ in range(n):
        a.append(int(data[ptr])); ptr += 1
    M = int(data[ptr]); ptr += 1

    result = two_unique_with_triples_mask(a, M)
    print(" ".join(map(str, result)))

if __name__ == "__main__":
    main()

File: dsa-problems/Bitwise/solutions/javascript/BIT-002-two-unique-with-triples-mask.js
const readline = require("readline");

class Solution {
  twoUniqueWithTriplesMask(a, M) {
    let splitBit = -1;

    // Step 1: Find valid split bit within Mask
    for (let i = 0; i < 31; i++) {
      if (((M >> i) & 1) === 0) continue;

      let count = 0;
      for (const x of a) {
        if ((x >> i) & 1) count++;
      }

      if (count % 3 === 1) {
        splitBit = i;
        break;
      }
    }

    let num1 = 0;
    let num2 = 0;

    // Step 2: Reconstruct using standard Single Number II logic per group
    for (let i = 0; i < 31; i++) {
      let c1 = 0;
      let c2 = 0;
      for (const x of a) {
        const bitVal = (x >> i) & 1;
        const group = (x >> splitBit) & 1;
        if (group === 1) {
          c2 += bitVal;
        } else {
          c1 += bitVal;
        }
      }
      if (c1 % 3 === 1) num1 |= 1 << i;
      if (c2 % 3 === 1) num2 |= 1 << i;
    }

    const result = [num1, num2];
    result.sort((x, y) => x - y);
    return result;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(line.trim()));
rl.on("close", () => {
  if (data.length === 0) return;
  const tokens = data.join(" ").split(/\s+/);
  if (tokens.length === 0 || tokens[0] === "") return;

  let ptr = 0;
  const n = Number(tokens[ptr++]);
  const a = [];
  for (let i = 0; i < n; i++) a.push(Number(tokens[ptr++]));
  const M = Number(tokens[ptr++]);

  const solution = new Solution();
  const result = solution.twoUniqueWithTriplesMask(a, M);
  console.log(result.join(" "));
});
